Upgrade to Angular 19

pawelfras opened the pull request:
- [x] Nx migration (all packages and their migration)

  - [x] ran `npx nx migrate latest` (bump most of libraries to the latest version) and `npx nx migrate --run-migration` (run packages' migration, all except angular's migration from initializer tokens to helper functions)
  - [x] bump '@angular/**' packages to version '^19.0.3'
  - [x] bump '@ng-select/ng-select' to version '^14.1.0'
  - [x] bump 'angular-oauth2-oidc' to version '19.0.0'
  - [x] bump 'ngx-infinite-scroll' to version '^19.0.0'
  - [x] bump 'zone.js' to version '0.15.0'
  - [x] bump '@angular-devkit/**' libraries to version '^19.0.4'
  - [x] bump '@angular-eslint/**' libraries to version '19.0.2'
  - [x] bump '@nx/**' libraries to version '^20.2.2'
  - [x] bump '@schematics/angular' to version '^19.0.4'
  - [x] bump 'jest-preset-angular' to version '14.4.2'
  - [x] bump 'ng-packagr' to version '^19.0.1'
  - [x] bump 'nx' to version '^20.2.2'
  - [x] bump '@ngrx/**' libraries to version `^19.0.0` 
  - [x] bump 'typescript' to version '^5.6.3'
  - [x] bump `@typescript-eslint` to `8.18.1`
  - [x] add `standalone: false` to all components, directives and pipes 
          for more, see: https://github.com/angular/angular/pull/57643
  - [x] add rule `"@angular-eslint/prefer-standalone": "off"` in eslint files
  - [x] remove deprecated `"angular-eslint/no-host-metadata-property"` 
          for more, see: https://github.com/angular-eslint/angular-eslint/pull/2113
  - [x] update '@angular/ssr' import paths to use the new '/node' entry point when 'CommonEngine' is detected
          for more, see: https://github.com/angular/angular-cli/pull/28278
- [x] Library `@angular-builders/custom-webpack`
  - [x] bump to version `19.0.0-beta.0`
  - [x] bump to stable `19.0.0` when ready
- [x] Unit tests
  - [x] update snapshots
  - [x] adjust ssr/engine tests to use new required `standalone: false` property
  - [x] adjust Jest config for `setup/ssr` library to use `beasties` from `node_modules` instead of the internalized one provided with `@angular/ssr` due to issues with not supported ES format
    for more, see: https://github.com/angular/angular-cli/pull/28228
    and: https://github.com/angular/angular-cli/pull/28726
- [x] Schematics
  - [x] refactor `server.ts` to use `dist/server/index.server.html` and adjust schematics 
  - [x] https://github.com/SAP/spartacus/pull/19878
- [x] Migration and installation docs
  - [x] migration docs
    - [x] migration to Angular v18
    - [x] migration to Angular v19 
    - [x] changes required for `application` builder users
  
QA (done by author)
- [x] migration from SPA 6.x with SSR (Angular 15)
- [x] migration from the latest SPA 2211.x (Angular 17)
- [x] migration from the latest SPA 2211.x with SSR (Angular 17)
- [x] installation of SPA in fresh Angular 19 app
- [x] installation of SPA with SSR in fresh Angular 19 app

QA (done by reviewer)
- [x] migration from SPA 6.x with SSR (Angular 15)
- [x] migration from the latest SPA 2211.x (Angular 17)
- [x] migration from the SPA 2211.x with SSR (Angular 17)
- [x] installation of SPA in fresh Angular 19 app
- [x] installation of SPA with SSR in fresh Angular 19 app

pawelfras committed:
fix: Issues In Unit Tests

- fix type issues
- use `takeUntil` with destroy subject in `trending-searches.service.ts` as a more reliable cleanup method to properly handle cleanup
- remove redundant 'declatartion: true' from 'tsconfig.schematics.json' files,
  that caused issues in unit tests due to type errors in '@schematics/angular' library
  'Buffer is not a generic type'. The issue appeared because '*.d.ts' files are taken into consideration when running tests, maybe unnecessary
  for more, see: https://github.com/angular/angular-cli/blob/4db4dd4315fd8c31872bbf1e82e3414eea15ffef/goldens/public-api/angular_devkit/schematics/index.api.md\?plain\=1\#L498

pawelfras committed:
chore: Nx migration (all packages and their migration)

- ran `npx nx migrate latest` (bump most of libraries to the latest version) and `npx nx migrate --run-migration` (run packages' migration, all except angular's migration from initializer tokens to helper functions)
- bump '@angular/animations' to version '^19.0.3'
- bump '@angular/common' to version '^19.0.3'
- bump '@angular/compiler' to version '^19.0.3'
- bump '@angular/core' to version '^19.0.3'
- bump '@angular/forms' to version '^19.0.3'
- bump '@angular/platform-browser' to version '^19.0.3'
- bump '@angular/platform-browser-dynamic' to version '^19.0.3'
- bump '@angular/platform-server' to version '^19.0.3'
- bump '@angular/pwa' to version '^19.0.4'
- bump '@angular/router' to version '^19.0.3'
- bump '@angular/service-worker' to version '^19.0.3'
- bump '@angular/ssr' to version '^19.0.4'
- bump '@ng-select/ng-select' to version '^14.1.0'
- bump 'angular-oauth2-oidc' to version '19.0.0'
- bump 'ngx-infinite-scroll' to version '^19.0.0'
- bump 'zone.js' to version '0.15.0'
- bump '@angular-devkit/build-angular' to version '^19.0.4'
- bump '@angular-devkit/core' to version '^19.0.4'
- bump '@angular-devkit/schematics' to version '^19.0.4'
- bump '@angular-eslint/builder' to version '19.0.2'
- bump '@angular-eslint/eslint-plugin' to version '^19.0.2'
- bump '@angular-eslint/eslint-plugin-template' to version '^19.0.2'
- bump '@angular-eslint/schematics' to version '^19.0.2'
- bump '@angular-eslint/template-parser' to version '^19.0.2'
- bump '@angular-eslint/test-utils' to version '^19.0.2'
- bump '@angular-eslint/utils' to version '^19.0.2'
- bump '@angular/cli' to version '^19.0.4'
- bump '@angular/compiler-cli' to version '^19.0.3'
- bump '@angular/language-service' to version '^19.0.3'
- bump '@nx/angular' to version '^20.2.2'
- bump '@nx/devkit' to version '^20.2.2'
- bump '@nx/eslint-plugin' to version '^20.2.2'
- bump '@nx/jest' to version '^20.2.2'
- bump '@nx/workspace' to version '^20.2.2'
- bump '@schematics/angular' to version '^19.0.4'
- bump 'jest-preset-angular' to version '14.4.2'
- bump 'ng-packagr' to version '^19.0.1'
- bump 'nx' to version '^20.2.2'
- bump 'typescript' to version '^5.6.3'
- add `standalone: false` to all components, directives and pipes
  for more, see: https://github.com/angular/angular/pull/57643
- add rule `"@angular-eslint/prefer-standalone": "off"` in eslint files
- remove deprecated `"angular-eslint/no-host-metadata-property"`
  for more, see: https://github.com/angular-eslint/angular-eslint/pull/2113
- update '@angular/ssr' import paths to use the new '/node' entry point when 'CommonEngine' is detected
  for more, see: https://github.com/angular/angular-cli/pull/28278

pawelfras committed:
chore: Change Version Range Specifier For Typescript

- change range specifier for typescript library to meet Angular's requirement (>=5.5.0 and <5.7.0)

pawelfras committed:
fix: Build issues

- fix building libs and CSR by solvig type errors provided with TS 5.6

pawelfras committed:
chore: Update '@angular-builders/custom-webpack' to '19.0.0-beta.0'

- this update is required for SSR to work, because previous version use Angualr 18 API which is no longer part of @angular/core in v19.
  for more, see: https://github.com/angular/angular/pull/58891

pawelfras committed:
chore: Update dependencies

pawelfras committed:
fix: Fix issues with schematics and related unit tests

- update snapshots with new budget thresholds
  for more, see: https://github.com/angular/angular-cli/pull/28255
- update snapshots with new packages versions
- update schematics and related tests with the new `server.ts` location
  for more, see: https://github.com/angular/angular-cli/pull/28571
- adjust positions in utils tests to reflect the `standalone: false` in component decorators
- adjust SSR schematics to remove `withEventReplay()` that Angular installs by default
  for more, see:
- refactor tests for 2211.19 SSR migration to make them independent of Angular's schematics
  for more, see: https://github.com/angular/angular-cli/pull/28469

pawelfras committed:
chore: Fix linter issues

- move `@angular-eslint/prefer-standalone` to global eslint rules
- add missing `OnDestroy` implementation in cds library
- remove `@angular-eslint/no-host-metadata-property` from the config as it is no longer part of the package
 for more, see: https://github.com/angular-eslint/angular-eslint/pull/2113

pawelfras committed:
chore: Fix prettier issues

- add missing comma to all files where 'standalone: false' was added

pawelfras committed:
chore: Update NgRx to v19.0.0-beta.0

pawelfras committed:
chore: Upgrade `@typescript-eslint` packages to 8.18.1

- upgrade to version that supports TS >= 5.6.3
- add `@typescript-eslint/rule-tester` which was missing peer dependency of `@angular-eslint/test-utils`
- adjust import-utils.ts to meet new typing

pawelfras committed:
chore: Update package-lock file

pawelfras committed:
chore: Upgrade NgRx to stable version 19.0.0

- bump `@ngrx/*` libraries to `19.0.0`
- adjust failing unit tests to satisfy types

pawelfras committed:
chore: Update dependencies and package-lock file

pawelfras committed:
fix: Jest tests issues

- update snapshots
- adjust ssr/engine tests to use new required `standalone: false`
- adjust jest config for `setup/ssr` unit tests to use `beasties` from node modules instead of the internalized one provided with `@angular/ssr` due to issues with not supported ES format

pawelfras committed:
Trigger Build

pawelfras committed:
Trigger Build

pawelfras committed:
chore: Initial migration docs

pawelfras committed:
refactor: Adjust installation schematics for SSR to use `server/index.server.html`

- adjust server file
- update snapshot

pawelfras committed:
docs: Update migration docs

- update docs with information about changing path to the index file to be in line with the builder's output

pawelfras committed:
chore Upgrade '@angular-builders/custom-webpack' to version 19.0.0

github-actions[bot] committed:
Add license header

pawelfras committed:
Revert "Add license header"

This reverts commit 3fb0b3a0f3c20c97ca9696599bf1f059831ef584.

github-actions[bot] committed:
Add license header

pawelfras committed:
Revert "Add license header"

This reverts commit c304b3421bc60d1a330631b1c49eae5406558a10.

pawelfras committed:
Merge branch 'develop-next-major' into epic/upgrade-to-angular-19

Platonn commented on package.json:
```diff
@@ -246,7 +246,7 @@
     "ts-morph": "^23.0.0",
     "ts-node": "^10.6.0",
-    "typescript": "^5.6.3",
+    "typescript": "~5.6.3",
```
Nice move 👍
because TS is often bringing breaking changes their "minor" releases.

Platonn commented on projects/schematics/src/add-ssr/index.ts:
```diff
@@ -66,7 +66,7 @@ declare module '@schematics/angular/utility/workspace-models' {
     /**
      * Since ng17 it's no more "browserTarget" but "buildTarget" property
      */
-    buildTarget?: string;
```
[MINOR] (can be done anytime, even after release)
I only guess this workaround (whole lines 64-71) might be not needed anymore (because I only guess angular fixed it themselves). Let's try if we can remove this block;

Platonn commented on integration-libs/cds/src/trending-searches/trending-searches.service.spec.ts:
```diff
@@ -87,11 +86,11 @@ describe('TrendingSearchesService', () => {
-        result = searchPhrases;
+      .subscribe((result) => {
+        // Verify the result
+        expect(result).toEqual(mockSearchPhrases);
```
[MEDIUM]
In this PR we've added `expect()` inside rxjs subscription. Let's call `done()` callback inside the subscription, to avoid a possible false-positive test result, when the "observable under test" would not emit anything

pawelfras replied:
This test uses `fakeAsync` and if I'm not mistaken, `done()` callback is not required in that case (even not possible to be called).

Platonn replied:
Thank you, I wasn't aware of this limitation.
Unfortunately, it doesn't change our problem - the `expect` inside a subscription callback is prone to false-positive test result, when the observable doesn't emit.

What about moving the expect out from the subscription callback?
```diff
+   let searchPhrases;
    const subscription = service
      .getTrendingSearches()
      .subscribe((result) => {
-        expect(result).toEqual(mockSearchPhrases);
+        searchPhrases = result;
      });
    
+    // perhaps `tick()` might need to be called here, but not necessarily
    
+    expect(searchPhrases).toEqual(mockSearchPhrases);
```

Platonn commented on integration-libs/cds/src/trending-searches/trending-searches.service.ts:
```diff
@@ -80,7 +81,7 @@ export class TrendingSearchesService implements OnDestroy {
         return timer(0, POLL_INTERVAL).pipe(
           switchMap(() => this.fetchTrendingSearches(url)),
-          takeWhile(() => !this.destroy$.value)
+          takeUntil(this.destroy$)
```
[NITPICK]
alternatively, new Angular API could be used instead
https://angular.dev/api/core/rxjs-interop/takeUntilDestroyed

Platonn commented on projects/schematics/src/migrations/2211_19/update-ssr/update-ssr-files_spec.ts:
[QUESTION]
Isn't it strange we adapted the migration schematics for 2211_19 (that is compatible only with ng17) in a PR that bumps Angular to 19?

pawelfras replied:
It turned out that the test for migration schematics for 2211_19 strongly depended on the Angular schematics in the currently installed version. The problem was only discovered when I bumped the version up to Angular 19 due to changes from how the created application looked in v17. I adjusted the tests to be deterministic and independent of external schematics.

Platonn commented on projects/schematics/src/shared/utils/file-utils_spec.ts:
```diff
@@ -432,28 +432,25 @@ describe('File utils', () => {
-          'AppComponent',
-          change
-        );
+        const testChange = new ReplaceChange(filePath, 192, original, change);
```
[QUESTION]
Just to better understand: why we had to change those original numbers in this file? 173, 188, 158, 171. Is it really related to this PR?

pawelfras replied:
Yes, it is - change is required due to the `standalone: false` in every component/directive/pipe decorator

Platonn commented on docs/migration/2211_ng19/migration.md:
```diff
@@ -2,11 +2,11 @@
 
 Before upgrading Spartacus to the new version with Angular 18, you need to first:
 - upgrade to the latest 2211.x of Spartacus
```
[MINOR] (can be changed later, but before doc freeze)
I guess it should be:
```suggestion
- upgrade Spartacus to version 2211.32.1 <or whatever latest version will be available _just before_ Feb release>
```

Platonn commented on docs/migration/2211_ng19/migration.md:
````diff
@@ -17,22 +17,22 @@ git commit -m "update angular 18 and 3rd party deps angular 18 compatible"
 ```
 
-### Adjust Angular configuration
+Note: Unselect `use-application-builder` migration when migrating to Angular 19. Applications created before SPA 2211.19 doesn't support this builder. Applications created starting from 2211.19 already supports it.
````
[MEDIUM]
This is said too late. People had already run the command. 
Instead let's put a bold warning _before running the command_ (in line 23). Also I'd mention that you need to hit a SPACE bar to unselect this option, and only then enter.

Because it's easy to miss the small prompt during the installation, perhaps we might even attach 2 screenshots:
- how the default preselected option looks like
- how the option unselected with a SPACE bar looks like

Platonn commented on docs/migration/2211_ng19/migration.md:
```diff
@@ -29,10 +29,25 @@ git commit -m "update angular 19 and 3rd party deps angular 19 compatible"
+
+### If using Server Side Rendering (SSR) and `application` builder
+
+For applications using `application` builder having SSR support, you need to adjust the `server.ts` file to be compatible with the output generated by the builder.
```
[MEDIUM]
The phrase "`application` builder" is a short term that we know, but many customers might not know. So let's be explicit what it means in practice - the value in `angular.json`. Feel free to choose a different wording than the suggested one below:
```suggestion
For applications with SSR support that use the Angular's `application` builder (i.e. having in `angular.json` the following value: `... "architect": { "build": { "builder": "@angular-devkit/build-angular:application", ...`), you need to adjust the `server.ts` file to be compatible with the output generated by this builder.
```

Platonn reviewed (dismissed):
I approve the PR, "code-wise".
But 1 comment marked as `[MEDIUM]` IMHO should be fixed at least before the final release.

The PR needs QA yet, that Florent kindly agreed to do. Thank you, Florent. 🙌 

When Florent approves "QA-wise", I'm fine with merging it to `develop-next-major`.

pawelfras committed:
chore: Change typescript version from '~5.6.3' to '^5.6.3'

- extend supported version of typescript for compatibility with new Angular versions

pawelfras committed:
Merge branch 'develop-next-major' into epic/upgrade-to-angular-19

pawelfras committed:
Merge branch 'develop-next-major' into epic/upgrade-to-angular-19

pawelfras committed:
Merge branch 'develop-next-major' into epic/upgrade-to-angular-19

pawelfras committed:
chore: move migration docs to folder '2211_ng19'

pawelfras committed:
Merge branch 'develop' into epic/upgrade-to-angular-19

pawelfras committed:
Merge branch 'develop' into epic/upgrade-to-angular-19

pawelfras committed:
Merge branch 'develop' into epic/upgrade-to-angular-19

pawelfras committed:
chore: refactor after review

pawelfras committed:
chore: Bump '@types/node' to v22.10.7

- bump `@types/node` library
- adjust information about compatible engines

pawelfras committed:
Merge branch 'develop' into epic/upgrade-to-angular-19

Platonn commented on docs/migration/2211_ng19/migration.md:
```diff
@@ -1,24 +1,37 @@
-- upgrade to the latest 2211.x of Spartacus
+
+- upgrade Spartacus to version 2211.32.1
+- install Node 22 version and upgrade `@types/node` to version 22
```
Those are 2 different big steps, lets not mix them in one bullet (so customers don't skip one of them unconsciously):
```suggestion
- install Node 22 version
- upgrade package `@types/node` to version 22
```

Platonn commented on docs/migration/2211_ng19/migration.md:
```diff
@@ -1,24 +1,37 @@
 
+> **Warning**
+>
+> Do not select `use-application-builder` migration when migrating to Angular 18. Applications created before SPA 2211.19 doesn't support this builder. Applications ?created starting from 2211.19 already supports it.
```
typos
```suggestion
> Do not select `use-application-builder` migration when migrating to Angular 18. Applications created before SPA 2211.19 doesn't support this builder. Applications created starting from 2211.19 already support it.
```

pawelfras committed:
chore: refactor after review

pawelfras committed:
Merge branch 'develop' into epic/upgrade-to-angular-19

Platonn committed:
feat: silence the deprecation warnings about Sass `@import`usage in new apps via our installation schematics (#19878)

fixes https://jira.tools.sap/browse/CXSPA-9272

**QA steps:**
1. on this branch `npm run build:libs`
2. on this branch publish the built libs to verdaccio (e.g. via `npx ts-node schematics/testing.ts`)
4. in a separate folder create a fresh ng19 app `npx @angular/cli@19 new test-app --standalone=false`
5. enter the subfolder of the test-app and there install spartacus (with libs from verdaccio) `ng add @spartacus/schematics@2211.31.0-1`
6. run the app `npm run start` and wait for the successful build
7. verify that the build output in the terminald doensn't contain warnings about Sass `@import` usage

BEFORE:
![image](https://github.com/user-attachments/assets/e456ce4e-05c8-404e-9670-c4178c09ef2f)

AFTER:
![image](https://github.com/user-attachments/assets/abd6fd77-2c35-4c8a-940a-39aafb54fc14)

pawelfras committed:
chore: adjust config for installation script

pawelfras committed:
Revert "chore: adjust config for installation script"

This reverts commit 1c4d8f97fcef076040cfb3618f5bb885a23c02f2.

pawelfras committed:
chore: adjust config for installation script

- use 'develop' branch
- add comment to switch to Angular 19 when PR with Angualr 19 is ready

pawelfras committed:
Merge branch 'develop' into epic/upgrade-to-angular-19
